### PR TITLE
POC for behavioral component Responsive

### PR DESCRIPTION
**Pull request by asci**
# What does this PR do:

This is Proof-of-concept implementation of `Responsive` component for UI kit. If you like it — give a thumb up 👍  I'll cover it with tests and we'll merge it.

The usage is:
```js

<Responsive gt='small'>
  I am not small
</Responsive>

<Responsive only='large'>
  Show only for large screens
</Responsive>
```

I also consider supporting children as a function where I'll pass current screen size or so, but no examples out of my mind for such case.

# Where should the reviewer start:

Diff and Living style guide

### CODE REVIEW

**Comment by fahad19** [comment]
Few questions:

* ~Are we following `travix-breakpoints` for this?~ Yes
* What happens if we try to server-side render these components?
* Global usage of `window`?

File: components/responsive/responsive.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import sizes from 'travix-breakpoints/sizes';
+import debounce from 'lodash.debounce';
```
Comment (fahad19) [inline components/responsive/responsive.js]: better do `lodash/debounce`. will play well along with our `externals` configuration in Webpack.

**Comment by asci** [comment]
2. For SSR, now it is not ready. We'll need to updates some other components as well
3. Yep, any other way to add a listener (I'll try with `global` but not sure it is supported everywhere)?

File: components/responsive/responsive.js (hunk 1 above)
Comment (asci) [reply components/responsive/responsive.js]: oke

**Comment by fahad19** [comment]
Using globals would be a big issue with SSR, so I would highly recommend staying away from them as much as possible.

---

I will never suggest to think from Frint's perspective, and I indeed want UI-Kit to stay its own independent project.

But it could have helped access `window` in different environments (server and client) quite well:

```js
import { createApp } from 'frint';

const App = createApp({
  providers: [
    {
      name: 'window',
      useFactory() {
        if (process.env.BABEL_ENV !== 'server') {
          return window; // I am in browser
        }

        // I am in a non-browser environment
       return fakeWindow;
      }
    }
  ]
});
```

In your components, you can then access `window` as:

```js
import { observe } from 'frint-react';

function MyComponent(props) {
  const { window } = props;

  return <p></p>;
}

export default observe(function (app) {
  return {
    window: app.get('window'),
  };
})(MyComponent);
```

Same code for accessing `window` in Components, without involving globals.

**Comment by asci** [comment]
I don't think we need to add new dependency here. We can safely check is this running in browser or SSR.

**Comment by asci** [comment]
so, for SSR we can define a default screen width. I offer to go with `tiny.max` — in this case, SSR Responsive will render for mobile devices, which are usually slower and no rerender is required.

**Comment by mAiNiNfEcTiOn** [comment]
😄 
### About this PoC...
Again, the breakpoints 🙄 . Now, declarative, via markup.

We've discussed this before, and as discussed before, I do feel that breakpoints in general should not be part of UI Kit.

With that said, let's go into the analysis:

1. 2 new dependencies: `lodash`, `travix-breakpoints`.
1. Users now are forced to use our breakpoints (if we use the responsive component in our elements).
1. Also, contrarily to what is common, I'm not going to provide you an example of code... But you can. Feel free to do so and test complex cases to. I can already see duplicated code all over the place. It looks like an AB Test wrapper, where "if A, render this", "If B, render that"... A `render` function that would have one single case, styled later at CSS level, becomes series of blocks of JSX to satisfy people who don't like to look or read CSS.

Contrarily to what I feel like doing, I won't decline/close/request changes to this PR. I will let y'all do what you want. But __above everything__ I'd like the @Travix-International/frontend and @Travix-International/frontend-epam to look at this and do a REAL review. Give your feedback, don't just 👍  or 👎... __Say why you like it and why you think it is important!__ Because it's part of your job to maintain and develop for this.

And after that do the same for the other PRs. Take your time. Even when stuff is closed and merged. Refute, discuss, ASK WHY! That's the whole purpose of discussing, the knowledge sharing! 😉 😄 

### Side note:
As this advances, your "global" or "high level" concerns are becoming _everything_ __but__ fixing the things necessary for the UI Kit to be in a stable version. So many elements missing, so many rules/contracts/interfaces to define. But no, _breakpoints_ are the most important thing. 😉

**Comment by asci** [comment]
Thanks @mAiNiNfEcTiOn 

**About feelings**
I feel breakpoints should be part of UI kit. You feel they shouldn’t. We’re both not wrong here. Because it’s about feelings. In order to solve this dilemma, we should have a clear purpose of UI kit. Probably some kind of manifesto aligned with designers and business. Do we develop abstract UI kit or we develop it for our company needs? My vision based on the presentation I did a year ago about UI kit. If something changed since then, I’m sorry, I was not aware of it.

**Dependencies**
Would it be better if I implement `debounce` by myself? We already have it available on the website. About `travix-breakpoints` — see above about purpose.

**DRY and CSS**
Even so, this component looks like AB test wrapper it has a different purpose, a different behavior and a different place to stay. I don’t want to prematurely create an abstraction here because bad abstraction could lead to more problems then duplicated code. And about doing responsive with CSS, 2 points here: 
	1) developers still will be able to use CSS media queries if the component will not fit their needs; 
	2) Doing responsiveness with JSX has a benefit: what you rendered with React will be actually rendered in a browser. With media queries, you need to render everything, keep it in DOM and then hide what should be hidden.

**Review**
I fully agree that we need to review PRs better. Ask questions, give feedback. It is really important. Would really like to hear from you guys, do you already see a place where this component will help?

**About side note**
I definitely agree that there is something more important to do. Why then I grab breakpoints? Few reasons. The main reason that I can develop it autonomously, without bothering design department and most of the developers (only for review). Secondly, breakpoints, as well as icons and grids, are still big missing parts and it is unclear how they should look like. We discussed it a couple of times, but without any visible output. So, here it is, we can discuss particular implementation for breakpoints.

Would be also nice to have a list of issues need to be solved (apart from few issues on Github)

Again thanks for raising important questions and providing interesting thoughts.

**Comment by fahad19** [comment]
what if, UI-Kit is a monorepo with these two packages:

```js
├── travix-ui-components/ (current ui-kit)
└── travix-ui-css/
```

The `travix-ui-css` can be an extremely small CSS framework, defining grid, responsiveness etc. I am thinking of something similar to http://tachyons.io/

The `travix-ui-components` package can then use the CSS framework. The components exported will be responsive by default, and not be required to be wrapped in any additional parent/wrapper components.

Those using UI-Kit, would then be asked to load both the JS and CSS bundles together (like other UI-Kits do currently, including https://ant.design/):

```html
<link rel="stylesheet" href="ui-kit.css" />
<script src="ui-kit.js"></script>
```

**Comment by iwwwi** [comment]
There are other question that concern me other that @mAiNiNfEcTiOn has mentioned.

@asci you say: 

> developers still will be able to use CSS media queries if the component will not fit their needs;

- my concern is that with this approach we will end up with developers using this component + css to define their needs for the css. :/ Which brings us to the chaotic distributed code between JSX & CSS

Other than that:
- what about the cases when I want to render my components for both xs && xlarge (same content) but not for the rest of the breakpoints.

**Comment by asci** [comment]
@fahad19
Splitting UI-kit is an interesting idea. And at some point, it looks quite natural. Keeping it in monorepo will remove the need to commit to different repos ( the pain we have now). From the other side having separate CSS framework will lead to more work without critical need — we will need to define and maintain contracts, class names, etc.

At least, for now, we can use monorepo for themes, that is a good idea, thanks, Fahad!

@iwwwi
1. I agree that it is not good, and I try to do everything to make developers not writing CSS :D The problem is I can’t deny writing own CSS with a component. There still will be an opportunity to provide custom styles\queries.
2. It is possible with current component, but it doesn’t look pretty:
```
const myAwesomeMarkup = this.renderSomeStuff();

<Responsive only=’tiny’>
  {myAwesomeMarkup}
</Responsive>

<Responsive only=‘xlarge'>
  {myAwesomeMarkup}
</Responsive>
```
I can make property `only` support an array, so it will look like this:
```
<Responsive only=[‘xlarge’, ’tiny’]>
  {myAwesomeMarkup}
</Responsive>
```
Would it solve the issue? 

Also, what do you think about static methods, like we have now in `Breakpoints` helper in the main codebase? We can simply extract it from there.

And thanks for raising good questions related to component API. This is something that would be good to have for most components we develop.

@yurist38
Good points, Yuri! The default theme is poor because nobody invests time in it… because there is not much business value in it. But I agree that we need to change it and at some point, we will do it.

**Comment by yurist38** [comment]
@asci I think it's quite important to make property `only` supporting an array of values. Otherwise, if we go the current way we will get DOM rendered twice, right?

**Comment by asci** [comment]
@iwwwi  @yurist38  Updated to support array with `only` prop. If no more comments, I'll cover with tests

File: components/responsive/responsive.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    if (!frameSize) {
+      console.warn(`Responsive: Passed unknown frame size name: ${only || gt || lt}. Ignoring any window width`);
+      frameSize = ignoreFrameSize;
+    }
+
+    if ((gt && lt) || (gt && only) || (lt && only)) {
+      console.warn('Responsive: Passed more than one frame size. `only` has a priority');
+    }
+    const state = {
+      width: isBrowser ? global.window.innerWidth : sizesNumbers.tiny.max - 1,
```
[the remark this answers is not in the document]
Comment (asci) [reply components/responsive/responsive.js]: nope, I'll use tiny screen size as default for SSR, so there should be minus

File: components/responsive/responsive.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      state.min = frameSize.min;
+      state.max = frameSize.max;
+    }
+
+    return state;
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = Responsive.getState(props);
```
[the remark this answers is not in the document]
Comment (asci) [reply components/responsive/responsive.js]: what is the benefit? I thought to move it out of the public available scope

File: components/responsive/responsive.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  componentDidMount() {
+    this.bindEvents();
+  }
+
+  componentWillUnmount() {
+    this.unbindEvents();
+  }
+
+  bindEvents = () => {
+    global.window.addEventListener('resize', this.handleResize);
```
[the remark this answers is not in the document]
Comment (asci) [reply components/responsive/responsive.js]: `bindEvents` will be called only when `componentDidMount` which will happen only in the browser

File: components/responsive/responsive.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        'large',
+        'xlarge',
+      ])),
+    ]),
+
+    /**
+     * **Warning**: If you render a string or multiple elements you need a wrapper element, like `div` or `span`.
+     * Keep it until React 16, when we can return fragments
+     */
+    wrap: PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.func]),
```
[the remark this answers is not in the document]
Comment (asci) [reply components/responsive/responsive.js]: why not? what is the type of a stateful component?

File: components/responsive/responsive.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      PropTypes.string,
+      PropTypes.element,
+      PropTypes.node,
+    ]).isRequired,
+
+    /**
+     * Show if screen size is greater then
+     */
+    gt: PropTypes.oneOf([
+      'tiny',
```
[the remark this answers is not in the document]
Comment (asci) [reply components/responsive/responsive.js]: than `react-styleguidist` will not catch it

File: components/responsive/responsive.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (reaktivo) [reply components/responsive/responsive.js]: `sizesNumbers.tiny.max - 1 == 318`, `sizesNumbers.small.min - 1  == 320`

File: components/responsive/responsive.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (reaktivo) [reply components/responsive/responsive.js]: I meant to specify when running on a test environment, but it's possible that `jsdom` already handles this case

File: components/responsive/responsive.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (reaktivo) [reply components/responsive/responsive.js]: The type of stateful component is still a function. Passing `<Response wrap={AnotherComponent} />` will work, while `<Response wrap={<AnotherComponent />} />` (which is a valid `PropTypes.element`) will throw, at least if [this lines](https://github.com/Travix-International/travix-ui-kit/pull/221/files#diff-a2214d71b55781c6b19ec17df80218e4R171) are not changed.

File: components/responsive/responsive.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (asci) [reply components/responsive/responsive.js]: but what if there will be gap 😂 , 
anyway, if you insist, I'll update

File: components/responsive/responsive.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (asci) [reply components/responsive/responsive.js]: got it, thanks!

**Comment by leandrooriente** [comment]
@asci Is this PR still valid? 

We have some cases in Seat Map widget where this solution could help us a lot.

**Comment by asci** [comment]
@leandrooriente  it is valid, but required few adjustments:
- fix conflicts
- fix comments from @reaktivo 
- remove `wrap` property and use React.Fragment instead
- get approves from @iwwwi @froskie @AlleeX @rbardini 
- merge
- ...
- PROFIT!

**Comment by leandrooriente** [comment]
Cool @asci, if you don't mind, I can help you to fix those problems

**Comment by asci** [comment]
@leandrooriente I would appreciate it, since I don't have time ATM

**Comment by froskie** [comment]
Since the debates for the grid/breakpoints will follow and a possible implementation is coming, this component will get outdated really soon. So we'll postpone its implementation to a nearby future.

**Pull request closed without merging**